msgpack-core/src/main/java/org/msgpack/core/buffer/MessageBufferU.java

```diff
@@ -0,0 +1,183 @@
+                resetBufferPosition();
+            }
+        }
+    }
+
+    private void resetBufferPosition() {
+        reference.position(0);
+        reference.limit(size);
+    }
+
+    @Override
+    public byte getByte(int index) {
+        return reference.get(index);
+    }
+    @Override
+    public boolean getBoolean(int index) {
+        return reference.get(index) != 0;
+    }
+    @Override
+    public short getShort(int index) {
+        return reference.getShort(index);
+    }
+    @Override
+    public int getInt(int index) {
+        return reference.getInt(index);
+    }
+    @Override
+    public float getFloat(int index) {
+        return reference.getFloat(index);
+    }
+    @Override
+    public long getLong(int index) {
+        return reference.getLong(index);
+    }
+    @Override
+    public double getDouble(int index) {
+        return reference.getDouble(index);
+    }
+    @Override
+    public void getBytes(int index, int len, ByteBuffer dst) {
+        try {
+            reference.position(index);
+            reference.limit(index+len);
+            dst.put(reference);
+        }
+        finally {
+            resetBufferPosition();
+        }
+    }
+    @Override
+    public void putByte(int index, byte v) {
+        reference.put(index, v);
+    }
+    @Override
+    public void putBoolean(int index, boolean v) {
+        reference.put(index, v ? (byte) 1 : (byte) 0);
+    }
+    @Override
+    public void putShort(int index, short v) {
+        reference.putShort(index, v);
+    }
+    @Override
+    public void putInt(int index, int v) {
+        reference.putInt(index, v);
+    }
+    @Override
+    public void putFloat(int index, float v) {
+        reference.putFloat(index, v);
+    }
+    @Override
+    public void putLong(int index, long l) {
+        reference.putLong(index, l);
+    }
+    @Override
+    public void putDouble(int index, double v) {
+        reference.putDouble(index, v);
+    }
+    @Override
+    public ByteBuffer toByteBuffer(int index, int length) {
+        try {
+            reference.position(index);
+            reference.limit(index+length);
+            return reference.slice();
+        }
+        finally {
+            resetBufferPosition();
+        }
+    }
+    @Override
+    public ByteBuffer toByteBuffer() {
+        return toByteBuffer(0, size);
+    }
+
+    @Override
+    public void getBytes(int index, byte[] dst, int dstOffset, int length) {
+        try {
+            reference.position(index);
+            reference.get(dst, dstOffset, length);
+        }
+        finally {
+            resetBufferPosition();
+        }
+    }
+
+    @Override
+    public void putByteBuffer(int index, ByteBuffer src, int len) {
+        assert (len <= src.remaining());
+
+        if(src.hasArray()) {
+            byte[] srcArray = src.array();
+            putBytes(index, srcArray, src.position(), len);
+
+        } else {
+            for(int i = 0; i < len; ++i) {
```

Comment:
`putByte()` with loop sounds slow (I saw it was around 5x slower). How about using `java.nio.ByteBuffer#put(java.nio.ByteBuffer)` and `java.nio.Buffer#limit(int)` like this?

```
            int currLimitOfSrc = src.limit();
            src.limit(src.position() + len);
            reference.put(src);
            src.limit(currLimitOfSrc);
```